### ISSUE TITLE
TTL Stored procedure stopped working

### ISSUE THREAD

**Issue by rohankharwar** [neo4j-contrib/neo4j-apoc-procedures#255]
Hi 

We are using APOC TTL to delete the nodes that are older than 90 days. 

However we noticed that suddenly the APOC queries for TTL stopped working after a specific time. 
I have been looking at Debug log and it does not indicate or give more information. I don't believe the Master instance failed and switched nodes either. 

2017-01-18 15:45:19.724+0000 INFO  526 ms: embedded-session	- MATCH (t:TTL) where t.ttl < timestamp() WITH t LIMIT 1000 DETACH DELETE t - {} - {}

2017-01-18 15:50:19.627+0000 INFO  429 ms: embedded-session	- MATCH (t:TTL) where t.ttl < timestamp() WITH t LIMIT 1000 DETACH DELETE t - {} - {}

2017-01-18 15:55:19.585+0000 INFO  387 ms: embedded-session	- MATCH (t:TTL) where t.ttl < timestamp() WITH t LIMIT 1000 DETACH DELETE t - {} - {}

We see the following queries in the query.log related to TTL.
But after 15:55 we did not see any TTL queries run. 

So wanted to check how to debug this or is there some issue.
Thanks and appreciate your help as always.
Regards,
Rohan

**Comment by jexp** [comment]
Hmm hard to figure out why. I could add some more logging while catching exceptions. But it should be rescheduled even after.

**Comment by rohankharwar** [comment]
Hi @jexp,

Definitely need more logging on this. 
I am grepping the query.log as 

`cat query.log | grep -i embedded-session`

```
2017-01-30 15:30:18.769+0000 INFO  531 ms: embedded-session	 - MATCH (t:TTL) where t.ttl < timestamp() WITH t LIMIT 1000 DETACH DELETE t - {} - {}
2017-01-30 15:35:18.543+0000 INFO  305 ms: embedded-session	 - MATCH (t:TTL) where t.ttl < timestamp() WITH t LIMIT 1000 DETACH DELETE t - {} - {}
2017-01-30 15:40:18.607+0000 INFO  369 ms: embedded-session	 - MATCH (t:TTL) where t.ttl < timestamp() WITH t LIMIT 1000 DETACH DELETE t - {} - {}
2017-01-30 15:45:18.604+0000 INFO  366 ms: embedded-session	 - MATCH (t:TTL) where t.ttl < timestamp() WITH t LIMIT 1000 DETACH DELETE t - {} - {}
```

So the query at 2017-01-30 15:45:18.604 was the last TTL query. I have the Debug.log and did not see any issues in the debug.log w.r.t failover to another instance or any such issue. 
There was only transaction log rotation and some checkpoints counts. 

Let us know if you have extra logging enabled we can test this out and figure out what is happening. 

Thanks,
Rohan

**Comment by tuddman** [comment]
@rohankharwar    were you able to isolate what's going on here?

was it that TTL _stopped_ working?  or was TTL - expiring nodes - not working _at all_ ?

I have `apoc.ttl.enabled=true`   set and  `apoc.ttl.frequency=30` set in neo4j.conf

running in two separate neo4j instances: one is `3.1.0` and the other is `3.0.7`

the `3.1.0` instance  has  `apoc-3.1.0.3-all.jar` installed

the `3.0.7` instance  has  `apoc-3.0.8.4-all.jar` installed

**both** instances have nodes with `:TTL` labels and `ttl` properties of timestamps that are in the past.

**neither** instance is deleting the expired nodes.

debug.log nor neo4j.log is showing any hints of anything related to TTL.

any suggestions on how to isolate / address / fix?

TL;DR - can't get TTL to work on `3.1.0` _nor_ `3.0.7` 


cc @jexp

**Comment by tuddman** [comment]
did [this](https://github.com/neo4j-contrib/neo4j-apoc-procedures/commit/e75da3ae9313c11bb1aa8988a7f62ff875e77cb9) commit   cause [this line](https://github.com/neo4j-contrib/neo4j-apoc-procedures/blob/3.1/src/main/java/apoc/ApocKernelExtensionFactory.java#L58) to stop functioning properly?

haven't worked out exactly the interplay between kernel scheduler and apoc scheduler.

**Comment by rohankharwar** [comment]
@tuddman : Sorry for the late reply. TTL would stop working. It would work fine. As part of TTL we were deleting few nodes that have expired.  Lets say we restart neo4j and then it would work fine. Then few hours later it would stop working. So we would not see the TTL queries in the debug.log.

So currently working with Michael to implement more logging so we can see why it stops working. 
Not much info to debug why it stops working.

**Comment by jexp** [comment]
@rohankharwar did you check again after the release ?

**Comment by jexp** [comment]
@rohankharwar ping
